### ISSUE TITLE
torch.scatter_ returns incorrect result when running on CPU using int64 indexes

### ISSUE THREAD

**Issue by leourbina** [pytorch/pytorch#39565]
## 🐛 Bug

When working on CPU, `scatter_` returns the wrong result (fails to raise an error, as in the GPU) when the indexes used are of type `torch.int64`

## To Reproduce

```python
x = torch.arange(1,5, device='cpu', dtype=torch.int64).reshape(2, 2)
torch.zeros(2, 2, device=device).scatter_(0, torch.tensor([[1, 0 ], [0, 0, ]], device=device), x)
```
Outputs:

```python
tensor([[2.8026e-45, 4.2039e-45],
        [1.4013e-45, 0.0000e+00]])
```

## Expected behavior

The correct behavior should be to raise an error as it seems like `int64` is not supported for indexing in this operation (although I'm not quite sure as to why not). When running on the GPU, we get the following:

```python
x = torch.arange(1,5, device='cuda:0', dtype=torch.int64).reshape(2, 2)
torch.zeros(2, 2, device=device).scatter_(0, torch.tensor([[1, 0 ], [0, 0, ]], device=device), x)
```
Raises:

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-53-8af60a187cc2> in <module>()
      1 device = 'cuda:0'
      2 x = torch.arange(1,5, device=device, dtype=torch.int64).reshape(2, 2)
----> 3 torch.zeros(2, 2, device=device).scatter_(0, torch.tensor([[1, 0 ], [0, 0, ]], device=device), x)

RuntimeError: Expected object of scalar type Float but got scalar type Long for argument #4 'src' in call to _th_scatter_
```

When switching the type of the index array to `float` we get the expected result (both on CPU and GPU):

```python
x = torch.arange(1,5, dtype=torch.float).reshape(2, 2)
torch.zeros(2, 2, device=device).scatter_(0, torch.tensor([[1, 0 ], [0, 0, ]], device=device), x)
```
Which outputs:
```
tensor([[3., 4.],
        [1., 0.]])
```

## Environment

PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect

cc @ezyang @gchanan @zou3519

**Comment by VitalyFedyunin** [comment]
Could be fixed already, need to check nightly.

**Comment by ngimel** [comment]
This is fixed on master, see #38646

**Comment by leowu2021** [comment]
Dtype via torch.zeros is torch.float32
